## Peer-graded Assignment: Capstone Project - The Battle of Neighborhoods (Week 1)

## Week 1 Prompt:

1.	Clearly define a problem or an idea of your choice, where you would need to leverage the Foursquare location data to solve or execute. Remember that data science problems always target an audience and are meant to help a group of stakeholders solve a problem, so make sure that you explicitly describe your audience and why they would care about your problem.

2. This submission will eventually become your Introduction/Business Problem section in your final report. I recommend that you push the report (having your Introduction/Business Problem section only for now) to your Github repository and submit a link to it.

### Discussion of the problem:

After considering the data available, I have decided to consider the following problem: In New York City (NYC), if someone is looking to open a restaurant, where would you recommend that they open it? A recommendation will be made at the 'neighborhood-level'.

A client or a group of restaurateurs would be interested in this project because setting up at a good starting point and greatly improve the chances of their businesses surviving and/or even thriving in the competitive industry today.

### Discussion of the background:

Since just simply reproducing what has already been taught in this course provides minimal value, I have searched the internet for more datasets that would be useful (discussed in more detail in the section on data). Besides looking at the number of existing potential competitors (as done in the course), I will also be using two additional datasets:

1) The historic NYPD shooting incident data

2) The population of each of the neighborhoods in each of the Borough in NYC (in years 2000 and 2010)

### Reasons for the additional datasets and the additional value they provide:

1) The number of shooting incidents in a neighborhood is a telling sign of the stability and level of violence/ criminality in that particular neighborhood, and should also definitely be of concern to any business owner.

2) However, merely comparing the number of shooting incidents across the neighborhoods would be an unfair comparison, since one can expect the number of shooting incidents to be greater at neighborhoods with a greater population size.

3) Therefore, the data on the population size of each neighborhood would be used to calculate the number of shooting incidents per capita for each of these neighborhoods. 

4) Additionally, the population data will be useful in estimating the % year-on-year population growth. However, since only the population of two years can be found in the data, the % year-on-year population growth would be calculated using these two years. The neighborhoods would be ranked from best to worst based on their the % year-on-year population growth. A high the % year-on-year population growth in a particular neighborhood is good for business, and may influence the decision on which neighborhood the business should be opened. 

### Data acquisition and usage

### 1) Locations of neighborhoods

The procedures to do so follows what has been covered in the online lectures, and are shown below:

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [2]:
with open('nyu_2451_34572-geojson.json') as json_data:
    newyork_data = json.load(json_data)

In [3]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [4]:
neighborhoods_data = newyork_data['features']

In [5]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [6]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [7]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

In [8]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [9]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [10]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]))

The dataframe has 5 boroughs and 306 neighborhoods.


However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [11]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data

Borough         Neighborhood   Latitude  Longitude
0   Manhattan          Marble Hill  40.876551 -73.910660
1   Manhattan            Chinatown  40.715618 -73.994279
2   Manhattan   Washington Heights  40.851903 -73.936900
3   Manhattan               Inwood  40.867684 -73.921210
4   Manhattan     Hamilton Heights  40.823604 -73.949688
5   Manhattan       Manhattanville  40.816934 -73.957385
6   Manhattan       Central Harlem  40.815976 -73.943211
7   Manhattan          East Harlem  40.792249 -73.944182
8   Manhattan      Upper East Side  40.775639 -73.960508
9   Manhattan            Yorkville  40.775930 -73.947118
10  Manhattan           Lenox Hill  40.768113 -73.958860
11  Manhattan     Roosevelt Island  40.762160 -73.949168
12  Manhattan      Upper West Side  40.787658 -73.977059
13  Manhattan       Lincoln Square  40.773529 -73.985338
14  Manhattan              Clinton  40.759101 -73.996119
15  Manhattan              Midtown  40.754691 -73.981669
16  Manhattan          Murray Hill  40.748303 -73.978332
17  Manhattan              Chelsea  40.744035 -74.003116
18  Manhattan    Greenwich Village  40.726933 -73.999914
19  Manhattan         East Village  40.727847 -73.982226
20  Manhattan      Lower East Side  40.717807 -73.980890
21  Manhattan              Tribeca  40.721522 -74.010683
22  Manhattan         Little Italy  40.719324 -73.997305
23  Manhattan                 Soho  40.722184 -74.000657
24  Manhattan         West Village  40.734434 -74.006180
25  Manhattan     Manhattan Valley  40.797307 -73.964286
26  Manhattan  Morningside Heights  40.808000 -73.963896
27  Manhattan             Gramercy  40.737210 -73.981376
28  Manhattan    Battery Park City  40.711932 -74.016869
29  Manhattan   Financial District  40.707107 -74.010665
30  Manhattan        Carnegie Hill  40.782683 -73.953256
31  Manhattan                 Noho  40.723259 -73.988434
32  Manhattan         Civic Center  40.715229 -74.005415
33  Manhattan        Midtown South  40.748510 -73.988713
34  Manhattan         Sutton Place  40.760280 -73.963556
35  Manhattan           Turtle Bay  40.752042 -73.967708
36  Manhattan           Tudor City  40.746917 -73.971219
37  Manhattan      Stuyvesant Town  40.731000 -73.974052
38  Manhattan             Flatiron  40.739673 -73.990947
39  Manhattan         Hudson Yards  40.756658 -74.000111

In [12]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


In [13]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

### 2) The number of venues nearby each neighborhood location (uses Foursquare API)

The procedures to do so follows what has been covered in the online lectures, and are shown below:

#### Define Foursquare Credentials and Version

In [14]:
CLIENT_ID = # 'your-client-ID' # your Foursquare ID
CLIENT_SECRET = # 'your-client-secret' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

# print('Your credentails:')
# print('CLIENT_ID: ' + CLIENT_ID)
# print('CLIENT_SECRET:' + CLIENT_SECRET)

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    LIMIT = 100 # limit of number of venues returned by Foursquare API
    radius = 500 # define radius
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

In [16]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


#### Let's check the size of the resulting dataframe

In [17]:
print(manhattan_venues.shape)
manhattan_venues.head()

(2996, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066      Starbucks   
4  Marble Hill              40.876551               -73.91066        Dunkin'   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877531       -73.905582    Coffee Shop  
4       40.877136       -73.906666     Donut Shop

Let's check how many venues were returned for each neighborhood

In [18]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       59                      59     59   
Carnegie Hill                           85                      85     85   
Central Harlem                          45                      45     45   
Chelsea                                 97                      97     97   
Chinatown                              100                     100    100   
Civic Center                            87                      87     87   
Clinton                                100                     100    100   
East Harlem                             43                      43     43   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                                96                      96     96   
Gramercy                                69                      69     69   
Greenwich Village                      100                     100    100   
Hamilton Heights                        61                      61     61   
Hudson Yards                            53                      53     53   
Inwood                                  58                      58     58   
Lenox Hill                             100                     100    100   
Lincoln Square                          93                      93     93   
Little Italy                           100                     100    100   
Lower East Side                         40                      40     40   
Manhattan Valley                        42                      42     42   
Manhattanville                          43                      43     43   
Marble Hill                             26                      26     26   
Midtown                                100                     100    100   
Midtown South                           93                      93     93   
Morningside Heights                     39                      39     39   
Murray Hill                             77                      77     77   
Noho                                   100                     100    100   
Roosevelt Island                        24                      24     24   
Soho                                    71                      71     71   
Stuyvesant Town                         17                      17     17   
Sutton Place                            93                      93     93   
Tribeca                                 69                      69     69   
Tudor City                              73                      73     73   
Turtle Bay                             100                     100    100   
Upper East Side                         85                      85     85   
Upper West Side                         70                      70     70   
Washington Heights                      88                      88     88   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                59               59              59  
Carnegie Hill                    85               85              85  
Central Harlem                   45               45              45  
Chelsea                          97               97              97  
Chinatown                       100              100             100  
Civic Center                     87               87              87  
Clinton                         100              100             100  
East Harlem                      43               43              43  
East Village                    100              100             100  
Finan

#### Let's find out how many unique categories can be curated from all the returned venues

In [19]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 329 uniques categories.


In [20]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Baby Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Stadium  Beer Bar  Beer Garden  Beer Store  \
0                   0         0            0           0   
1                   0         0            0           0   
2                   0         0            0           0   
3                   0         0            0           0   
4                   0         0            0           0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  Building  \
0               0            0       0                0         0   
1               0    

And let's examine the new dataframe size.

In [21]:
manhattan_onehot.shape

(2996, 330)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [22]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000        0.000000               0.00   
1         Carnegie Hill           0.000000        0.000000               0.00   
2        Central Harlem           0.000000        0.000000               0.00   
3               Chelsea           0.000000        0.000000               0.00   
4             Chinatown           0.000000        0.000000               0.00   
5          Civic Center           0.000000        0.000000               0.00   
6               Clinton           0.000000        0.000000               0.00   
7           East Harlem           0.000000        0.000000               0.00   
8          East Village           0.000000        0.000000               0.00   
9    Financial District           0.000000        0.000000               0.00   
10             Flatiron           0.000000        0.000000               0.00   
11             Gramercy           0.000000        0.000000               0.00   
12    Greenwich Village           0.000000        0.000000               0.00   
13     Hamilton Heights           0.000000        0.000000               0.00   
14         Hudson Yards           0.000000        0.000000               0.00   
15               Inwood           0.000000        0.000000               0.00   
16           Lenox Hill           0.000000        0.000000               0.01   
17       Lincoln Square           0.000000        0.000000               0.00   
18         Little Italy           0.000000        0.000000               0.00   
19      Lower East Side           0.000000        0.000000               0.00   
20     Manhattan Valley           0.000000        0.000000               0.00   
21       Manhattanville           0.000000        0.000000               0.00   
22          Marble Hill           0.000000        0.000000               0.00   
23              Midtown           0.000000        0.000000               0.00   
24        Midtown South           0.000000        0.000000               0.00   
25  Morningside Heights           0.000000        0.000000               0.00   
26          Murray Hill           0.000000        0.000000               0.00   
27                 Noho           0.000000        0.000000               0.00   
28     Roosevelt Island           0.000000        0.000000               0.00   
29                 Soho           0.000000        0.000000               0.00   
30      Stuyvesant Town           0.000000        0.000000               0.00   
31         Sutton Place           0.000000        0.010753               0.00   
32              Tribeca           0.000000        0.000000               0.00   
33           Tudor City           0.000000        0.000000               0.00   
34           Turtle Bay           0.000000        0.000000               0.00   
35      Upper East Side           0.000000        0.000000               0.00   
36      Upper West Side           0.000000        0.000000               0.00   
37   Washington Heights           0.011364        0.000000               0.00   
38         West Village           0.000000        0.000000               0.00   
39            Yorkville           0.000000        0.000000               0.00   

    African Restaurant  American Restaurant  Antique Shop    Arcade  \
0             0.000000             0.000000      0.000000  0.000000   
1             0.000000             0.011765      0.000000  0.000000   
2             0.066667             0.044444      0.000000  0.000000   
3             0.000000             0.020619      0.000000  0.000000   
4             0.000000             0.030000      0.000000  0.000000   
5             0.000000             0.034483      0.011494  0.000000   
6             0.000000             0.030000      0.000000  0.000000   
7             0.000000             0.000000      0.000000  0.000000   
8             0.000000             0.010000      0

#### Let's confirm the new size

In [23]:
manhattan_grouped.shape

(40, 330)

#### Let's print each neighborhood along with the top 5 most common venues

In [24]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0           Park  0.10
1          Hotel  0.07
2  Memorial Site  0.05
3  Boat or Ferry  0.05
4            Gym  0.05


----Carnegie Hill----
           venue  freq
0    Coffee Shop  0.08
1    Yoga Studio  0.04
2      Bookstore  0.04
3    Pizza Place  0.04
4  Grocery Store  0.04


----Central Harlem----
                 venue  freq
0   African Restaurant  0.07
1                  Bar  0.04
2  American Restaurant  0.04
3   Seafood Restaurant  0.04
4   Chinese Restaurant  0.04


----Chelsea----
                venue  freq
0         Art Gallery  0.16
1         Coffee Shop  0.06
2      Ice Cream Shop  0.03
3  Italian Restaurant  0.03
4           Juice Bar  0.02


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.09
1               Bakery  0.04
2         Cocktail Bar  0.04
3                  Spa  0.03
4  American Restaurant  0.03


----Civic Center----
               venue  freq
0        Coffee Shop  0.06
1                S

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [26]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park                 Hotel   
1      Carnegie Hill           Coffee Shop           Yoga Studio   
2     Central Harlem    African Restaurant    Chinese Restaurant   
3            Chelsea           Art Gallery           Coffee Shop   
4          Chinatown    Chinese Restaurant                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0                   Gym         Boat or Ferry         Memorial Site  
1             Wine Shop           Pizza Place   Japanese Restaurant  
2     French Restaurant        Cosmetics Shop    Seafood Restaurant  
3    Italian Restaurant        Ice Cream Shop                 Hotel  
4          Cocktail Bar          Optical Shop    Salon / Barbershop

### 3) NYPD Shooting Incident Data (Historic)

In [27]:
from sodapy import Socrata
import pickle

In [28]:
# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:
client = Socrata("data.cityofnewyork.us", None)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.cityofnewyork.us,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# First 20000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
results = client.get("833y-fsy8", limit=20000)

# Convert to pandas DataFrame
incident_df = pd.DataFrame.from_records(results)

# Check shape
incident_df.shape

(20000, 18)

In [29]:
# extract out the data we are interested in:
incident_df = incident_df[incident_df.boro == 'MANHATTAN']
incident_df = incident_df[['boro','latitude','longitude']]

In [30]:
incident_df.head()

boro           latitude           longitude
1   MANHATTAN  40.80024432600004  -73.95339008999997
8   MANHATTAN  40.79415025600008  -73.93986908699996
27  MANHATTAN  40.80186520800004  -73.95723931799995
30  MANHATTAN  40.80941319900006  -73.94436716399997
33  MANHATTAN  40.79501283900004  -73.93613752499994

In [31]:
import folium
from folium import plugins
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [32]:
shooting_map = folium.Map(location=[latitude, longitude], zoom_start=11.5, prefer_canvas = True)

# mark each station as a point
for index, row in incident_df.iterrows():
    folium.Circle([row['latitude'], row['longitude']],
                        radius=15,
                        fill_color="#3db7e4", # divvy color
                       ).add_to(shooting_map)
    
# convert to (n, 2) nd-array format for heatmap
incidentArr = incident_df[['latitude', 'longitude']].values

# plot heatmap
shooting_map.add_child(plugins.HeatMap(incidentArr))

shooting_map

Besides storing the NYPD Shooting Incident Data into a dataframe, a simple algorithm will also be written to assign each shooting incident to a neighborhood; based on the latitude-longitude of the neighborhood location the incident is closest to.

### 4) Next, we acquire the population by neighborhood

In [33]:
# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:
client = Socrata("data.cityofnewyork.us", None)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.cityofnewyork.us,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# First 20000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
results = client.get("swpk-hqdp", limit=390)

# Convert to pandas DataFrame
population_df = pd.DataFrame.from_records(results)

# Check shape
population_df.shape

(390, 6)

In [34]:
population_df.head()

borough fips_county_code nta_code                         nta_name  \
0   Bronx              005     BX01               Claremont-Bathgate   
1   Bronx              005     BX03  Eastchester-Edenwald-Baychester   
2   Bronx              005     BX05       Bedford Park-Fordham North   
3   Bronx              005     BX06                          Belmont   
4   Bronx              005     BX07                        Bronxdale   

  population  year  
0      28149  2000  
1      35422  2000  
2      55329  2000  
3      25967  2000  
4      34309  2000

In [35]:
# Extract the data we are interested in:
population_df = population_df[population_df.borough == 'Manhattan']
population_df = population_df[population_df.year == '2010']
population_df = population_df[['borough','nta_name','population','year']]

In [36]:
population_df.head()

borough                           nta_name population  year
284  Manhattan                 Marble Hill-Inwood      46746  2010
285  Manhattan  Central Harlem North-Polo Grounds      75282  2010
286  Manhattan                   Hamilton Heights      48520  2010
288  Manhattan                     Manhattanville      22950  2010
289  Manhattan                Morningside Heights      55929  2010

In [37]:
population_df.shape

(29, 4)

In [38]:
temp_nta = population_df['nta_name'].str.split('-').apply(pd.Series, 1).stack()
temp_nta.index = temp_nta.index.droplevel(-1) # to line up with df's index
temp_nta.name = 'nta_name' # needs a name to join
del population_df['nta_name']
population_df = population_df.join(temp_nta)
population_df.columns = ['Borough','Population','Year','Neighborhood']

In [39]:
population_df.shape

(47, 4)

In [40]:
population_df.head()

Borough Population  Year          Neighborhood
284  Manhattan      46746  2010           Marble Hill
284  Manhattan      46746  2010                Inwood
285  Manhattan      75282  2010  Central Harlem North
285  Manhattan      75282  2010          Polo Grounds
286  Manhattan      48520  2010      Hamilton Heights

In [41]:
add_dataframe = pd.merge(manhattan_data, population_df, how='left', on=['Borough','Neighborhood'])

In [42]:
add_dataframe.dropna(how='any', thresh=None, subset=None, inplace=True)
add_dataframe.reset_index(drop=True, inplace=True)
add_dataframe.drop(columns=['Year'], inplace = True)
add_dataframe

Borough         Neighborhood   Latitude  Longitude Population
0   Manhattan          Marble Hill  40.876551 -73.910660      46746
1   Manhattan            Chinatown  40.715618 -73.994279      47844
2   Manhattan               Inwood  40.867684 -73.921210      46746
3   Manhattan     Hamilton Heights  40.823604 -73.949688      48520
4   Manhattan       Manhattanville  40.816934 -73.957385      22950
5   Manhattan      Upper East Side  40.775639 -73.960508      61207
6   Manhattan            Yorkville  40.775930 -73.947118      77942
7   Manhattan           Lenox Hill  40.768113 -73.958860      80771
8   Manhattan     Roosevelt Island  40.762160 -73.949168      80771
9   Manhattan      Upper West Side  40.787658 -73.977059     132378
10  Manhattan       Lincoln Square  40.773529 -73.985338      61489
11  Manhattan              Clinton  40.759101 -73.996119      45884
12  Manhattan              Midtown  40.754691 -73.981669      28630
13  Manhattan          Murray Hill  40.748303 -73.978332      50742
14  Manhattan              Chelsea  40.744035 -74.003116      70150
15  Manhattan         East Village  40.727847 -73.982226      44136
16  Manhattan      Lower East Side  40.717807 -73.980890      72957
17  Manhattan         Little Italy  40.719324 -73.997305      42742
18  Manhattan         West Village  40.734434 -74.006180      66880
19  Manhattan  Morningside Heights  40.808000 -73.963896      55929
20  Manhattan             Gramercy  40.737210 -73.981376      27988
21  Manhattan    Battery Park City  40.711932 -74.016869      39699
22  Manhattan        Carnegie Hill  40.782683 -73.953256      61207
23  Manhattan         Civic Center  40.715229 -74.005415      42742
24  Manhattan        Midtown South  40.748510 -73.988713      28630
25  Manhattan           Turtle Bay  40.752042 -73.967708      51231
26  Manhattan      Stuyvesant Town  40.731000 -73.974052      21049
27  Manhattan         Hudson Yards  40.756658 -74.000111      70150

In [43]:
# create map of Manhattan using latitude and longitude values
manhattan_population = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, neighborhood, population in zip(add_dataframe['Latitude'], 
                                              add_dataframe['Longitude'], 
                                              add_dataframe['Neighborhood'], 
                                              add_dataframe['Population']):
    
    label = '{}, {}'.format(neighborhood, population)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(manhattan_population)  
    
manhattan_population